### PR TITLE
feat: Realtime HTTP APIs

### PR DESCRIPTION
**Pull request by himank**
Staging on top of this PR: https://github.com/tigrisdata/tigris/pull/692

This diff implements HTTP APIs for real-time functionality. The HTTP APIs that we would be supporting will be

- GetRTChannel: To get information about the real-time channel
- GetRTChannels: To get information about all the real-time channels
- ReadMessages: To read messages from a position, the default would be read from now().
- Messages: To publish messages to a channel
- ListSubscriptions: A subscription list includes devices that are attached to the channel.

The real-time will be running on its own container using the same binary as the server but the server type will be realtime.

### CODE REVIEW

File: server/services/v1/realtime.go [hunk 1]
```diff
@@ -103,35 +110,81 @@ func (s *realtimeService) DeviceConnectionHandler(w http.ResponseWriter, r *http
...
 		return
 	}
 
-	log.Debug().Msgf("params '%v'", params)
-	_ = conn.WriteMessage(websocket.TextMessage, []byte(`{"error": "not implemented"}`))
-	_ = conn.Close()
+	ctx := r.Context()
+	session, err := s.devices.AddDevice(ctx, conn, params)
+	if err != nil {
+		realtime.SendReply(conn, params.ToEncodingType(), api.EventType_error, errors.InternalWS(err.Error()))
```
Comment (garrensmith) [inline server/services/v1/realtime.go:120]: This needs to be a RealtimeMessage type otherwise the client cannot decode it.

Comment (himank) [reply server/services/v1/realtime.go:120]: this is updated in another diff, will rebase now.

File: server/services/v1/realtime.go [hunk 2]
```diff
@@ -136,21 +138,53 @@ func (s *realtimeService) Ping(_ context.Context, _ *api.HeartbeatEvent) (*api.H
...
-}
+	runner := s.rtmRunner.GetChannelRunner()
+	runner.SetListSubscriptionsReq(req)
+
+	resp, err := s.devices.ExecuteRunner(ctx, runner)
+	if err != nil {
+		return nil, err
+	}
+	return resp.Response.(*api.ListSubscriptionResponse), nil
+}
```
Comment (JigarJoshi) [inline server/services/v1/realtime.go:190]: nit: newline missing.

File: server/services/v1/realtime/runner.go [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+			return nil, errors.NotFound("channel '%s' not present ", runner.channelReq.Channel)
+		}
+
+		return &Response{
+			Response: &api.GetRTChannelResponse{
+				Channel: channels[0],
+			},
+		}, nil
+	}
+}
```
Comment (JigarJoshi) [inline server/services/v1/realtime/runner.go:281]: nit: missing new line.

File: server/services/v1/service.go [hunk 4]
```diff
@@ -41,7 +41,6 @@ func GetRegisteredServicesRealtime(kvStore kv.KeyValueStore, searchStore search.
 	var v1Services []Service
 	v1Services = append(v1Services, newRealtimeService(kvStore, searchStore, tenantMgr, txMgr))
 	v1Services = append(v1Services, newHealthService(txMgr))
-	v1Services = append(v1Services, newObservabilityService(tenantMgr))
```
Comment (JigarJoshi) [inline server/services/v1/service.go:44]: why is this removed?

File: server/muxer/http.go [hunk 5]
```diff
@@ -36,26 +36,34 @@ type HTTPServer struct {
...
+		Inproc: &inprocgrpc.Channel{},
+		Router: chi.NewRouter(),
+	}
 
-	r.Use(cors.AllowAll().Handler)
-	r.Mount("/debug", chi_middleware.Profiler())
+	server.SetupMiddlewares(cfg)
 
+	// mount debug handler after adding all middlewares
+	server.Router.Mount("/debug", chi_middleware.Profiler())
```
Comment (JigarJoshi) [inline server/muxer/http.go:47]: can we mark this as an admin endpoint.

File: server/muxer/http.go [hunk 6]
```diff
@@ -36,26 +36,34 @@ type HTTPServer struct {
...
-	match := mux.Match(cmux.HTTP1Fast())
+	match := mux.Match(cmux.HTTP1Fast(), cmux.HTTP1HeaderField("Upgrade", "websocket"))
 	go func() {
 		srv := &http.Server{Handler: s.Router, ReadHeaderTimeout: readHeaderTimeout}
 		err := srv.Serve(match)
 		log.Fatal().Err(err).Msg("start http server")
 	}()
 	return nil
-}
+}
```
Comment (JigarJoshi) [inline server/muxer/http.go:69]: same.

File: server/services/v1/realtime/sessions.go [hunk 7]
```diff
@@ -111,3 +113,16 @@ func (s *Sessions) AddDevice(ctx context.Context, conn *websocket.Conn, params C
...
+	if err != nil {
+		return nil, err
+	}
+	tenant, err := s.tenantMgr.GetTenant(ctx, namespaceForThisSession)
+	if err != nil {
+		return nil, errors.NotFound("tenant '%s' not found", namespaceForThisSession)
+	}
+
+	return runner.Run(ctx, tenant)
+}
```
Comment (JigarJoshi) [inline server/services/v1/realtime/sessions.go:128]: same

File: server/services/v1/service.go (hunk 4 above)
Comment (himank) [reply server/services/v1/service.go:44]: I don't think we need it for real-time. But probably for console we need it?

**Review by JigarJoshi — APPROVED** [review]
